src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelPostProcessor.java

```diff
@@ -0,0 +1,1145 @@
+
+        /* store sample data verbatim */
+        this.sampleNames = sampleNames;
+        this.sampleTargets = sampleTargets;
+        this.sampleForwardBackwardResults = sampleForwardBackwardResults;
+        this.sampleBestPaths = sampleBestPaths;
+
+        /* the list of alleles must be in the same order as the corresponding lists of hidden states */
+        allAlleles = allStates.stream()
+                .map(AlleleMetadataProvider::toAllele)
+                .collect(Collectors.toList());
+        alternativeAlleles = alternativeStates.stream()
+                .map(AlleleMetadataProvider::toAllele)
+                .collect(Collectors.toList());
+
+        /* pool targets from all samples and sort them lexicographically */
+        final Set<T> pooledTargetsSet = new LinkedHashSet<>();
+        IntStream.range(0, numSamples)
+                .forEach(si -> pooledTargetsSet.addAll(sampleTargets.get(si).targets()));
+        masterTargetCollection = new HashedListTargetCollection<>(new ArrayList<>(pooledTargetsSet).stream()
+                .sorted(IntervalUtils.LEXICOGRAPHICAL_ORDER_COMPARATOR).collect(Collectors.toList()));
+    }
+
+    /***********************
+     * main public methods *
+     ***********************/
+
+    /**
+     * Write segments to a table writer
+     *
+     * @param writer an instance of {@link HiddenStateSegmentRecordWriter}
+     */
+    public void writeSegmentsToTableWriter(@Nonnull final HiddenStateSegmentRecordWriter<S, T> writer) {
+        if (allSegmentsBySampleName == null) {
+            performSegmentation();
+        }
+        final Iterator<HiddenStateSegmentRecord<S, T>> allSegmentRecordsSortedByCoordinates =
+                composeTargetSortedSegmentRecordIterator(masterTargetCollection, allSegmentsBySampleName);
+        logger.info("Writing segments to file...");
+        writeSegments(allSegmentRecordsSortedByCoordinates, writer);
+    }
+
+    /**
+     * Create the VCF file based on segmentation of the provided input data
+     *
+     * @param outputWriter variant context writer
+     * @param variantPrefix a prefix to attach to variant IDs
+     * @param commandLine (optional) command line used to generate the data handled to this class
+     */
+    public void writeVariantsToVCFWriter(@Nonnull final VariantContextWriter outputWriter,
+                                         @Nonnull final String variantPrefix,
+                                         @Nullable final String commandLine) {
+        if (allSegmentsBySampleName == null) {
+            performSegmentation();
+        }
+        logger.info("Generating genotyping segments...");
+        final List<GenotypingSegment> genotypingSegments = composeGenotypingSegments(allSegmentsBySampleName);
+        composeVariantContextAndWrite(genotypingSegments, outputWriter, variantPrefix, commandLine);
+    }
+
+    /**
+     * Create the VCF file based on pre-determined segments (though, using the forward-backward result
```

Comment:
I think the idea is job avoidance: one precomputes and gives the tool the segments (let's say, the combined output of germline CNV calls on 1'000'000'000'000 samples). The tool skips segmentation and only creates VCF; of course, in addition to segments, we must also save the forward-backward results to disk and load them on par with segments. Let's keep the method as is for now. I'll make an issue for that.